File: python/semanticabi/abi/item/ItemType.py
from enum import Enum


class ItemType(Enum):
    """
    The different possible values for the `type` field in an ABI item.

    @author garrett
    """

    EVENT = "event"
    FUNCTION = "function"
    CONSTRUCTOR = "constructor"
    FALLBACK = "fallback"
    RECEIVE = "receive"
    ERROR = "error"

    @staticmethod
    def is_function_type(item_type: 'ItemType') -> bool:
        """
        Returns true if the item type is a function type.
        """
        return item_type in [ItemType.FUNCTION, ItemType.CONSTRUCTOR, ItemType.FALLBACK, ItemType.RECEIVE]

File: js/src/semanticabi/abi/item/ItemType.ts
import {Enum} from "common/Enum";

export class ItemType extends Enum {

    static readonly EVENT = new this('event', false);
    static readonly FUNCTION = new this('function', true);
    static readonly CONSTRUCTOR = new this('constructor', true);
    static readonly FALLBACK = new this('fallback', true);
    static readonly RECEIVE = new this('receive', true);
    static readonly ERROR = new this('error', false);

    constructor(name: string, isFunction: boolean) {
        super(name);
    }

}
ItemType.finalize();